### PR TITLE
scp03: command and associated TA to provision scp03 keys

### PR DESCRIPTION
**Pull request by ldts**
Please do not submit a Pull Request via github.  Our project makes use of
mailing lists for patch submission and review.  For more details please
see https://www.denx.de/wiki/U-Boot/Patches

### CODE REVIEW

File: cmd/Kconfig [hunk 1]
```diff
@@ -2183,6 +2183,16 @@ config CMD_FIOVB
...
 	    fiovb write_pvalue - write to RPMB
 	    fiovb read_pvalue - read from RPMB
+
+config CMD_SCP03
+	bool "scp03 - SCP03 enable and rotate/provision operations"
+	depends on SCP03
+	default y
+	help
+          Enables the SCP03 commands to activate I2C channel encryption and
+	  provision the SCP03 keys (on boards in the secure state)
```
Comment (ricardosalveti) [inline cmd/Kconfig:2193]: Mix of tabs x spaces.

File: include/scp03.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/*
+ * (C) Copyright 2021, Foundries.IO
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#ifndef	_SCP03_H
```
Comment (ricardosalveti) [inline include/scp03.h]: Tab x space.

**Comment by ricardosalveti** [comment]
Can you name this with OP-TEE/TEE or SE050 since this is specific to this setup? Generic SCP03 is confusing.

File: common/Kconfig [hunk 3]
```diff
@@ -964,6 +964,14 @@ config SPL_FIOVB
 	 This option enables compilation of Foundries.IO verified boot
 	 operations
 
+config SCP03
+	bool "Build SCP03 support"
+	depends on PARTITION_UUIDS
```
Comment (igoropaniuk) [inline common/Kconfig]: is `PARTITION_UUIDS` really needed?

File: cmd/Kconfig [hunk 4]
```diff
@@ -2183,6 +2183,16 @@ config CMD_FIOVB
 	    fiovb write_rb - write rollback index
 	    fiovb write_pvalue - write to RPMB
 	    fiovb read_pvalue - read from RPMB
+
+config CMD_SCP03
```
Comment (igoropaniuk) [inline cmd/Kconfig:2187]: I guess the first thing that you'll be asked while upstreaming that (if it's planned) - to cover it with python tests :)

File: cmd/scp03.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+ * (C) Copyright 2021, Foundries.IO
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
```
Comment (igoropaniuk) [inline cmd/scp03.c]: minor: SPDX-License-Identifier before copyright:

```
// SPDX-License-Identifier: GPL-2.0+                                            
/*                                                                              
 * (C) Copyright ...                                      
 */         
```

File: include/scp03.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/*
+ * (C) Copyright 2021, Foundries.IO
```
Comment (igoropaniuk) [inline include/scp03.h:3]: same for SPDX

File: include/tee/optee_ta_scp03.h [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+#define PTA_SCP03_UUID { 0xbe0e5821, 0xe718 ,0x4f77, \
+			{ 0xab, 0x3e, 0x8e ,0x6c ,0x73 ,0xa9 ,0xc7, 0x35 } }
+
+/*
+ * Enable SCP03 support on the Secure Element.
+ *   Setting the operation value to something different than NULL will trigger
+ *   an SCP03 provisioning operation when the board is in the closed state.
+ *
+ * in	params[0].a = operation
+ *
```
Comment (igoropaniuk) [inline include/tee/optee_ta_scp03.h:16]: minor: extra empty line in comment

File: cmd/Kconfig (hunk 1 above)
Comment (igoropaniuk) [inline cmd/Kconfig:2193]: I would have added one-two sentences what is SCP03 and where it's used.

**Comment by ldts** [comment]
> Can you name this with OP-TEE/TEE or SE050 since this is specific to this setup? Generic SCP03 is confusing.

I just followed what we did for for FIOVB for consistency..I think we should merge as is and then move both TAs FIOVB and SCP03 to a TEE specific folder

File: include/scp03.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ */
+
+#ifndef _SCP03_H
+#define _SCP03_H
+
+#include <common.h>
+
+/* Requests to OPTEE to enable or provision the Secure Channel Protocol on its
```
Comment (igoropaniuk) [inline include/scp03.h]: minor:
```
/* 
  * Requests to OPTEE to enable or provision the Secure Channel Protocol on its
```

**Review by igoropaniuk — APPROVED** [review]

File: cmd/Kconfig (hunk 1 above)
Comment (ldts) [reply cmd/Kconfig:2193]: fixed

File: include/scp03.h (hunk 2 above)
Comment (ldts) [reply include/scp03.h]: fixed

File: common/Kconfig (hunk 3 above)
Comment (ldts) [reply common/Kconfig]: nope, removed

File: cmd/Kconfig (hunk 4 above)
Comment (ldts) [reply cmd/Kconfig:2187]: well, we will have to cross that bridge when they ask :)

File: cmd/scp03.c (hunk 5 above)
Comment (ldts) [reply cmd/scp03.c]: um ok, uboot seems to have a mismatch of these all over...

File: include/tee/optee_ta_scp03.h (hunk 7 above)
Comment (ldts) [reply include/tee/optee_ta_scp03.h:16]: ok

File: include/scp03.h (hunk 8 above)
Comment (ldts) [reply include/scp03.h]: ok. pushing an update.

File: common/Kconfig [hunk 9]
```diff
@@ -964,6 +964,14 @@ config SPL_FIOVB
 	 This option enables compilation of Foundries.IO verified boot
 	 operations
 
+config SCP03
+	bool "Build SCP03 - Secure Channel Protocol O3 - controls"
+	depends on OPTEE
```
Comment (ricardosalveti) [inline common/Kconfig:969]: It depends on both TEE and OPTEE (and on our current branch OPTEE won't force TEE, not sure if it is still the case upstream).

Without TEE it fails to build (tee_ functions missing).

File: common/Kconfig [hunk 10]
```diff
@@ -964,6 +964,14 @@ config SPL_FIOVB
...
+	default y
```
Comment (ricardosalveti) [inline common/Kconfig]: Should we really default to y? I would think this should be an opt-in feature instead.

**Review by ricardosalveti — CHANGES_REQUESTED** [review]
Comments inline.

File: common/Kconfig (hunk 10 above)
Comment (ldts) [reply common/Kconfig]: yes. sorry I took this shortcut but forgot to remove it (I thought this was an urgent PR so I kind of rushed it)...will remove

File: common/Kconfig (hunk 9 above)
Comment (ldts) [reply common/Kconfig:969]: I'll check that

**Comment by ldts** [comment]
@ricardosalveti updated

**Review by ricardosalveti — APPROVED** [review]
LGTM, thanks!

**Pull request merged**